[49582] Build the modal dialog to share a work package

HDinger opened the pull request:
- [x] Add a feature flag
- [x] Add a button "Share"
  - [x] In full view
  - [x] In split view
- [x] On Click, a dialog opens
  - [x] User autocompleter
    - [x] Remove necessity to have `manage_members` to see all users.
    - [x] When adding an already added user, update that share's role and place it on top 
  - [x] Role selector
  - [x] Invite Button
  - [x] List of already invited users
- [x] Adding new members
  - [x] permission check
  - [x] Email sending
- [x] Deleting existing members
  - [x] permission check
- [x] Updating existing members
  - [x] permission check
- [x] Change in data (add or delete) leads to refresh of the modal content
- [x] Permission check (UI)
  - [x] Add prevented
  - [x] Update prevented
  - [x] Delete prevented 
- [x] The UI is build with Primer components
- [x] Build shared work package mailer view
- [x] Add Tests

### Out of scope

**Button**
* Show number of shares inside the button in the WP full view

**Modal Content**
* Checkboxes
* Filter for Type
* Filter for Role
* Show roles below user name

**Autocompleter**
* Show only users which are not invited yet
* Inline invite via autocompleter
* Show current role of a user in the autocompleter results
* Multiselect
* Share with groups

### Work packages

* https://community.openproject.org/wp/50176
* https://community.openproject.org/wp/50175
* https://community.openproject.org/wp/49582

HDinger committed:
Add dummy share WP button which opens a dummy Modal

ulferts committed:
i18n for share button

ulferts committed:
initial feature spec for sharing

ulferts committed:
add wp#shares controller stub

HDinger committed:
Add first iteration of ShareModalComponent

ulferts committed:
turn work package roles into builtin ones

HDinger committed:
Extract permission button into own component

ulferts committed:
simplify scope

ulferts committed:
allow assigning wp roles now that they are builtin

ulferts committed:
display share modal content

ulferts committed:
introduce scope for fetching users having an entity membership

ulferts committed:
use plural for route

HDinger committed:
Add userAutocompleter inside a Form which can be submitted and then refreshes the modal content

ulferts committed:
adapt feature spec to also cover adding shares

ulferts committed:
use a shares resource for routing

HDinger committed:
Add loading indicator for shareWP modal

HDinger committed:
Show icon on ShareWP trigger button

HDinger committed:
Simplify attributes of InviteUserForm

HDinger committed:
Add real label for search input

HDinger committed:
Use correct labels and extract path to PathHelper

HDinger committed:
Replace loading animation with skeleton

HDinger committed:
Add share button to split view

HDinger committed:
Update permission button label when selecting a valua via the dynamic_label attribute

ulferts committed:
validation and default values on share creation

ulferts committed:
add share deletion

aaron-contreras committed:
Clean up require statements

Everything under /spec/support is required already in
`rails_helper.rb`.

aaron-contreras committed:
Annotate spec file with `frozen_string_literal: true` magic comment

aaron-contreras committed:
Enable cuprite in `share_spec.rb`

Aid in feature spec performance. ~2 seconds chopped off!

aaron-contreras committed:
Fix factory cascades in `share_spec.rb`

Since `:project` was not specified in the `:member` factories, a new project
was being created for every membership. This created unintended memberships
where they were all in separate projects (one created for each member factory)
instead of in the one the work package actually belongs to.

A `work_package_member` factory exists which is to be used in this scenario and
solves this precise problem where the project is not derived from the entity which
would result in an invalid membership.

As a side benefit, this speeds up test times by spending
less time in Factory creation.

**Before**
```
[TEST PROF INFO] Time spent in factories: 00:00.738 (7.18% of total time)
[TEST PROF INFO] Factories usage

 Total: 32
 Total top-level: 10
 Total time: 00:00.738 (out of 00:14.335)
 Total uniq factories: 11

   total   top-level     total time      time per call      top-level time               name

       8           0        0.0238s            0.0030s             0.0000s notification_setting
       8           7        0.1582s            0.0198s             0.1517s               user
       5           1        0.4119s            0.0824s             0.2531s            project
       4           0        0.1741s            0.0435s             0.0000s             member
       1           0        0.0136s            0.0136s             0.0000s             status
       1           0        0.0115s            0.0115s             0.0000s view_work_package_role
       1           0        0.0048s            0.0048s             0.0000s comment_work_package_role
       1           0        0.0071s            0.0071s             0.0000s edit_work_package_role
       1           1        0.0345s            0.0345s             0.0345s               role
       1           1        0.2990s            0.2990s             0.2990s       work_package
       1           0        0.0150s            0.0150s             0.0000s           priority
```

**After**
```
[TEST PROF INFO] Time spent in factories: 00:00.586 (5.77% of total time)
[TEST PROF INFO] Factories usage

 Total: 28
 Total top-level: 10
 Total time: 00:00.586 (out of 00:14.193)
 Total uniq factories: 11

   total   top-level     total time      time per call      top-level time               name

       8           0        0.0222s            0.0028s             0.0000s notification_setting
       8           7        0.1628s            0.0204s             0.1560s               user
       4           0        0.0127s            0.0032s             0.0000s work_package_member
       1           1        0.1280s            0.1280s             0.1280s       work_package
       1           0        0.0115s            0.0115s             0.0000s           priority
       1           0        0.0172s            0.0172s             0.0000s             status
       1           0        0.0110s            0.0110s             0.0000s view_work_package_role
       1           0        0.0043s            0.0043s             0.0000s comment_work_package_role
       1           0        0.0058s            0.0058s             0.0000s edit_work_package_role
       1           1        0.0297s            0.0297s             0.0297s               role
       1           1        0.2723s            0.2723s             0.2723s            project
```

aaron-contreras committed:
Generate starter sharing_mailer_spec file

Barebones test setup for the sharing mailer.

aaron-contreras committed:
Generate SharingMailer preview file

I'd like to have a live preview of the mailer as I'm developing it.

aaron-contreras committed:
Generate SharingMailer with the `shared_work_package` mailer views

* Runs `./bin/rails generate Sharing shared_work_package`

* Customizes mail addressing to the membership's principal.

* Customizes mail subject with a custom locale.

HDinger committed:
Allow only users to be selected for sharing

HDinger committed:
Add empty state for share WP modal

aaron-contreras committed:
Create shared_work_package mailer header

ulferts committed:
require less permissions for API endpoint

ulferts committed:
add access control to shares controller

ulferts committed:
choose the correct role when creating

ulferts committed:
handle role selection on invite

ulferts committed:
sort active user shares by last updated at

ulferts committed:
allow share_work_package permission to see all principals

aaron-contreras committed:
Create shared_work_package mail body

ulferts committed:
disable component name linting

ulferts committed:
guard button by feature flag

aaron-contreras committed:
Add Mail Headers to shared_work_package mail

ulferts committed:
linting

ulferts committed:
add rel to link

ulferts committed:
spec lacking share_work_packages permission

HDinger commented on app/views/sharing_mailer/shared_work_package.html.erb:
```diff
@@ -0,0 +1,37 @@
+  <tr>
+    <%= placeholder_cell('12px', vertical: true) %>
+    <td>
+      <%= render partial: 'sharing_mailer/notification_mailer_header',
```
Hi @aaron-contreras 

I'd like to avoid having to add all these extra templates (mailer_body, mailer_header and notification_row) as we have all of them already and its hard to maintain multiple ones.
I am aware that the current Figma mockups define the mails slightly different, so I talked to the product team again and they agreed, that the emails should look exactly like out existing ones. So there is no need to add additional templates anymore but you can just use the exisiting ones instead.

ulferts committed:
introduce wp_member update contract&service

HDinger committed:
Merge remote-tracking branch 'origin/dev' into implementation/49582-build-the-modal-dialog-to-manage-members-of-a-work-package-using-hotwire-+-stimulus

HDinger committed:
Apply grid_layout on the user row in shareWP modal

HDinger committed:
Cleanup component names and implement min-height for the shareWP modal

HDinger committed:
Add permission check in the UI for sharing WP

HDinger committed:
Append autocompleter dropdown to body to avoid the modal being too small to fit it in

ulferts committed:
update existing member when adding the user again

ulferts committed:
adapt spec to read only mode

aaron-contreras committed:
Rework email design to adapt to existing styles

aaron-contreras committed:
Remove unused i18n keys

aaron-contreras committed:
Add text-version of shared_work_package email

aaron-contreras committed:
Create `WorkPackageSharedJob` to enqueue emails

aaron-contreras committed:
Add `after_perform` hook to send notifications on successful shares

aaron-contreras committed:
Add Mail delivery expectations on `share_spec`

ulferts committed:
implement updating of share

ulferts committed:
simplify interface of invite_user_form_component

ulferts committed:
adapt copy project job to work with role_ids

Using role_ids is supported by the Members::UpdateService

ulferts committed:
Merge remote-tracking branch 'origin/dev' into implementation/49582-build-the-modal-dialog-to-manage-members-of-a-work-package-using-hotwire-+-stimulus

ulferts committed:
finalize test on non editable modal

aaron-contreras committed:
Fix: `component_wrapper` rendering

Use single `component_wrapper` with `turbo-frame` as the tag. This
keeps a single element with the wrapper_key ID and preventing an
invalid turbo stream response due to duplicate IDs.

ulferts committed:
alter default role

ulferts committed:
remove todo - will do error handling when the need arises

ulferts committed:
extract into scopes

ulferts committed:
Merge remote-tracking branch 'origin/dev' into implementation/49582-build-the-modal-dialog-to-manage-members-of-a-work-package-using-hotwire-+-stimulus

ulferts committed:
stabilize spec

ulferts committed:
still stabilizing spec

aaron-contreras commented on spec/features/work_packages/share_spec.rb:
```diff
@@ -125,10 +125,8 @@
-      share_modal.close
-
-      share_modal.expect_closed
+      # Reopening the modal will show the same state as before.
+      work_package_page.visit!
```
👍

ulferts committed:
use wrapper_key method

oliverguenther commented on app/components/work_packages/share/modal_body_component.html.erb:
```diff
@@ -0,0 +1,69 @@
+  # Todo: doubled ID
+  component_wrapper do
+  # The identifier is a magic key also found in the wp-share modal template.
+    turbo_frame_tag wrapper_key do
```
```suggestion
  # The identifier is a magic key also found in the wp-share modal template.
  component_wrapper('turbo-frame') do
```

component_wrapper is calling `wrapper_key` interall,y

aaron-contreras replied:
I had addressed this in a separate branch with this one as the base. I cherry-picked the commit onto this branch: https://github.com/opf/openproject/pull/13752/commits/6874062bce4d13f945ed9772c7b1a31b86e09346

oliverguenther commented on frontend/src/app/features/work-packages/openproject-work-packages.module.ts:
```diff
@@ -463,6 +468,7 @@ import { RecentItemsService } from 'core-app/core/recent-items.service';
     WorkPackageSplitViewComponent,
     BackButtonComponent,
   ],
+  schemas: [CUSTOM_ELEMENTS_SCHEMA],
```
Why is this part necessary? I'm not seeing a opce- being used inside the new templates, so I don't think this would be required.

ulferts replied:
It was necessary back then because we had some back and forth with the autocompleter but you are right that it no longer is for the current solution so I removed it again.

ulferts replied:
Turns out I was wrong and we still need that declaration [for the `turbo-frame` tag](https://github.com/opf/openproject/actions/runs/6416153569/job/17419335271). But it does feel bad to disable the check for all custom tags within the module if all we want is angular to ignore one. I would be interested in knowing if there is a better way.

oliverguenther reviewed (changes requested):
Some tiny code remarks

- [ ] Adding a new share doesn't seem to do anything

![Kapture 2023-10-04 at 15 04 28](https://github.com/opf/openproject/assets/459462/40b1090e-25e4-4e95-be20-ac06d93c3d7b)

The turbo_stream response is made, but the list is not updated.

aaron-contreras committed:
Fix: Scope the shares for the user to those for the current work package

In a scenario where a user would have been shared two work packages, the
current query did not guarantee that the first share is for the current
work package.

aaron-contreras committed:
Merge branch 'dev' into implementation/49582-build-the-modal-dialog-to-manage-members-of-a-work-package-using-hotwire-+-stimulus

ulferts committed:
remove no longer necessary custom element declaration

ulferts committed:
Revert "remove no longer necessary custom element declaration"

This reverts commit 65b1b177d6917931a4857a703914cf651a15a2db.

ulferts commented:
I am merging this PR now. 

We can still address the issues, @oliverguenther mentioned later on.
